ilp-core/src/main/java/org/interledger/core/InterledgerAddressPrefix.java

```diff
@@ -0,0 +1,298 @@
+   *
+   * @return An {@link InterledgerAddressPrefix} representing the root prefix for the supplied
+   *     Interledger address.
+   */
+  default InterledgerAddressPrefix getRootPrefix() {
+    return this.getPrefix()
+        .map(InterledgerAddressPrefix::getRootPrefix)
+        .orElse(this);
+  }
+
+  /**
+   * <p>An implementation of {@link InterledgerAddress} that enforces allowed value per
+   * RFC-15.</p>
+   *
+   * <p>This immutable is interned because it only holds a {@link String} value, which itself is
+   * interned via the Java String pool.</p>
+   */
+  @Value.Immutable
+  @Value.Style(
+      typeBuilder = "*Builder",
+      visibility = Value.Style.ImplementationVisibility.PRIVATE,
+      builderVisibility = Value.Style.BuilderVisibility.PUBLIC,
+      defaults = @Value.Immutable(intern = true))
+  abstract class AbstractInterledgerAddressPrefix implements InterledgerAddressPrefix {
+
+    private static final String VALID_ADDRESS_PREFIX_REGEX
+        = "(?=^.{1,1023}$)^(g|private|example|peer|self|test[1-3]?|local)([.][a-zA-Z0-9_~-]+)*$";
+    private static final Pattern VALID_ADDRESS_PATTERN = Pattern
+        .compile(VALID_ADDRESS_PREFIX_REGEX);
+
+    private static final int ADDRESS_MIN_SEGMENTS = 1;
+    private static final String SCHEME_REGEX = "(g|private|example|peer|self|test[1-3]?)";
+    private static final String SEGMENT_REGEX = "[a-zA-Z0-9_~-]+";
+    private static final String SEPARATOR_REGEX = "[.]";
+    private static final String ADDRESS_LENGTH_BOUNDARIES_REGEX = "(?=^.{1,1021}$)";
+
+    /**
+     * Precondition enforcer that ensures the value is a valid Interledger Address.
+     *
+     * @see "https://github.com/interledger/rfcs/blob/master/0015-ilp-addresses/0015-ilp-addresses.md"
+     */
+    @Value.Check
+    void check() {
+      if (!VALID_ADDRESS_PATTERN.matcher(getValue()).matches()) {
+        // For performance reasons, we only do we do deeper introspection of the error if the input
+        // fails the VALID_ADDRESS_PATTERN check.
+        throw new IllegalArgumentException(getFirstInvalidityCause(getValue()));
+      }
+    }
+
+    /**
+     * Inspect an invalid Address String and determine the cause.
+     *
+     * @param invalidAddressString A {@link String} containing an invalid Interledger Address.
+     *
+     * @return An error String.
+     */
+    private String getFirstInvalidityCause(final String invalidAddressString) {
+      // validate no trailing period.
+      if (invalidAddressString.endsWith(".")) {
+        return Error.ILLEGAL_ENDING.getMessageFormat();
+      }
+
+      final List<String> schemeAndSegments = Arrays.asList(
+          invalidAddressString.split(SEPARATOR_REGEX, -1)
+      );
+      // validates scheme prefix format
+      final String schemePrefix = schemeAndSegments.get(0);
+      if (!Pattern.compile(SCHEME_REGEX).matcher(schemePrefix).matches()) {
+        return String.format(Error.INVALID_SCHEME_PREFIX.getMessageFormat(), schemePrefix);
```

Comment:
No win in precompiling the regexes ?

Reply:
That's a great point - this code was part of an external submission in `InterledgerAddress` that I mostly just adapted here. I'll try that out.

Reply:
Fixed.